### PR TITLE
[1/x] Prepare code sprint and implement repository type detection 

### PR DESCRIPTION
**Pull request by felixsch**
> This is the revert of the revert of the original pull request

This PR is part of our code sprint: https://trello.com/c/hU27QHob/3134-rdeb-sprint-tracker

original pull request: https://github.com/SUSE/rmt/pull/1056
revert of the pull request: https://github.com/SUSE/rmt/pull/1061

**The whole changeset is:**

[[1/5] Prepare code sprint and implement repository type detection](https://github.com/SUSE/rmt/pull/1066)
[[2/5] Add flat debian mirroring support to RMT](https://github.com/SUSE/rmt/pull/1057)
[[3/5] Refactor suma mirroring](https://github.com/SUSE/rmt/pull/1059)
[[4/5] Refactor repomd mirroring](https://github.com/SUSE/rmt/pull/1060)
[[5/5] Fixes and cleanup](https://github.com/SUSE/rmt/pull/1067)
[[6/6] Nested debian repository support](https://github.com/SUSE/rmt/pull/1070)

**How to test this pull request:**

*Make sure mirroring is working as expected*
```bash
$ bin/rmt-cli mirror
# Expect: mirroring to work normally
```

*Test repository type detection*
```ruby
# Sync any of our test accounts or our RMT reference account
$ rails c
> debian = Repository.find_by(name: 'Debian-10-SUSE-Manager-Tools')
> repomd = Repository.find_by(name: 'SLE-Product-SLED15-SP6-Pool')
> debian_mirror = RMT::Mirror.new(repository: debian, logger: nil, base_dir: '/tmp')
> repomd_mirror = RMT::Mirror.new(repository: repomd, logger: nil, base_dir: '/tmp')


> repomd_mirror.detect_repository_type
# expect: to return :repomd

> debian_mirror.detect_repository_type
# expect: to return :debian
```

**DO NOT MERGE UNTIL COMPLETELY READY**

### CODE REVIEW

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/351595/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 9a3288a34463a97d250b532bef0874c0d34117dd into 4926fb902d3e9c4583c4251415585c8cad14c22f)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/352701/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 9a3288a34463a97d250b532bef0874c0d34117dd into 91457227b05f9b2e7b124f0934f2008e32a9c4e0)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/353355/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 71f4674ef39d2191219a59d573290dabf4e8fa96 into 91457227b05f9b2e7b124f0934f2008e32a9c4e0)

**Review by josegomezr — COMMENTED** [review]
Check out the open questions, the suggestions are non-binding :P

File: lib/rmt/mirror.rb [hunk 1]
```diff
@@ -1,205 +1,38 @@
...
+      # e.g.
+      # https://update.suse.com/SUSE/product/some-product
+      # becomes
+      # https://update.suse.com/SUSE/product/some-product?authenication_tokensiduhashasdyashdaysdasud
+      uri = URI.join(url)
+      uri.query = @repository.auth_token if @repository.auth_token
+
+      request = RMT::HttpRequest.new(uri, method: :head, followlocation: true)
+      request.on_success do
+        return key
```
[the remark this answers is not in the document]
Comment (ngetahun) [reply lib/rmt/mirror.rb:32]: I would say almost impossible. OBS would need to support distribution-agnostic repository structure, which I don't think is possible yet

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/354738/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 71f4674ef39d2191219a59d573290dabf4e8fa96 into 95024c6f8813925a891c00abe62835118d54f960)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/354752/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 71f4674ef39d2191219a59d573290dabf4e8fa96 into 024478b7c8a9b7c45d617377f2c472917220d603)

**Pull request merged**